File: neuralgentics-cli/src/neuralgentics/merge.py
"""``opencode.json`` merge algorithm (§4 of the init-cli design doc).

The merge is a *deep merge with user-preservation semantics*: the user's
existing ``opencode.json`` is the base, the shipped config is the overlay,
and the algorithm only ever ADDS missing entries — it never removes,
overwrites existing dicts, or touches the ``provider`` block.

Public API
----------
* :func:`merge_opencode_json` — pure merge, returns new dict.
* :func:`merge_opencode_json_with_diff` — merge + human-readable change list.
* :func:`format_diff_for_display` — pretty-print a change list.
* :func:`parse_opencode_json` — ``json.loads`` wrapper raising
  :class:`OpenCodeJsonInvalid`.
* :func:`serialize_opencode_json` — canonical ``json.dumps`` format for
  stable round-trips.
* :data:`PLUGIN_REFERENCE`, :data:`INSTRUCTIONS_REFERENCE` — the canonical
  entries added to the ``plugin`` / ``instructions`` arrays.
"""

from __future__ import annotations

import copy
import json
from collections.abc import Mapping
from typing import Any

from neuralgentics.errors import OpenCodeJsonInvalid

__all__ = [
    "INSTRUCTIONS_REFERENCE",
    "PLUGIN_REFERENCE",
    "format_diff_for_display",
    "merge_opencode_json",
    "merge_opencode_json_with_diff",
    "parse_opencode_json",
    "serialize_opencode_json",
]

#: The plugin entry added to the ``plugin`` array.
PLUGIN_REFERENCE = "@veedubin/neuralgentics"

#: The instructions entry added to the ``instructions`` array.
INSTRUCTIONS_REFERENCE = "AGENTS.md"

#: Top-level scalar keys added from the shipped config when missing.
_TOP_LEVEL_SCALARS: tuple[str, ...] = (
    "$schema",
    "autoupdate",
    "tool_output",
    "compaction",
    "small_model",
)

#: Dict-valued sections merged key-by-key (add new keys only).
_DICT_SECTIONS: tuple[str, ...] = ("mcp", "lsp", "formatter")


def _union_arrays(user: list[object] | None, shipped: list[object] | None) -> list[object]:
    """Case-sensitive, order-preserving union of two arrays.

    User entries come first (in their original order), followed by any
    shipped entries not already present. Duplicates are dropped.
    """
    result: list[object] = []
    seen: set[object] = set()
    for source in (user, shipped):
        if not source:
            continue
        for item in source:
            # Use a hashable key for membership tracking. Lists/dicts are
            # unhashable; fall back to id() for those (rare in opencode.json
            # arrays, which are normally strings).
            try:
                key = item
                hash(key)
            except TypeError:
                key = id(item)
            if key not in seen:
                seen.add(key)
                result.append(item)
    return result


def merge_opencode_json(user_json: dict[str, Any], shipped_json: dict[str, Any]) -> dict[str, Any]:
    """Merge a user ``opencode.json`` with the shipped config.

    Implements §4.1 of the design doc exactly. Returns a NEW dict; the
    inputs are never mutated.

    Rules:
      1. ``plugin`` array → union (dedup, case-sensitive).
      2. ``instructions`` array → union (dedup).
      3. ``provider`` → preserved entirely from the user.
      4. ``mcp`` / ``lsp`` / ``formatter`` → add shipped keys missing in user.
      5. Top-level scalars (``$schema``, ``autoupdate``, ``tool_output``,
         ``compaction``, ``small_model``) → add from shipped if missing.
    """
    result: dict[str, Any] = copy.deepcopy(user_json)

    # 1 & 2 — array sections.
    result["plugin"] = _union_arrays(user_json.get("plugin"), shipped_json.get("plugin"))
    result["instructions"] = _union_arrays(
        user_json.get("instructions"), shipped_json.get("instructions")
    )

    # 3 — provider: PRESERVE user's entirely when present. When the user has
    # no provider block (e.g. empty user_json / "no existing opencode.json"
    # edge case in §4.5), add the shipped one verbatim.
    if "provider" not in result and "provider" in shipped_json:
        result["provider"] = copy.deepcopy(shipped_json["provider"])

    # 4 — dict sections: add shipped keys missing in user.
    for section in _DICT_SECTIONS:
        shipped_section = shipped_json.get(section)
        if not isinstance(shipped_section, Mapping):
            continue
        result_section = result.setdefault(section, {})
        for key, value in shipped_section.items():
            if key not in result_section:
                result_section[key] = copy.deepcopy(value)

    # 5 — top-level scalars.
    for key in _TOP_LEVEL_SCALARS:
        if key not in result and key in shipped_json:
            result[key] = copy.deepcopy(shipped_json[key])

    return result


def merge_opencode_json_with_diff(
    user_json: dict[str, Any], shipped_json: dict[str, Any]
) -> tuple[dict[str, Any], list[str]]:
    """Like :func:`merge_opencode_json` but also returns change descriptions.

    The returned list is empty when nothing changed (idempotent re-run).
    Change strings are stable and human-readable, e.g.::

        Added '@veedubin/neuralgentics' to plugin array
        Added MCP server 'searxng'
        Set small_model
    """
    result = merge_opencode_json(user_json, shipped_json)
    changes: list[str] = []

    # Plugin additions.
    user_plugin = user_json.get("plugin") or []
    shipped_plugin = shipped_json.get("plugin") or []
    for item in shipped_plugin:
        if item not in user_plugin:
            changes.append(f"Added {item!r} to plugin array")

    # Instructions additions.
    user_instr = user_json.get("instructions") or []
    shipped_instr = shipped_json.get("instructions") or []
    for item in shipped_instr:
        if item not in user_instr:
            changes.append(f"Added {item!r} to instructions array")

    # Dict-section additions.
    section_labels = {"mcp": "MCP server", "lsp": "LSP server", "formatter": "formatter"}
    for section, label in section_labels.items():
        shipped_section = shipped_json.get(section)
        if not isinstance(shipped_section, Mapping):
            continue
        user_section = user_json.get(section)
        if not isinstance(user_section, Mapping):
            user_section = {}
        for key in shipped_section:
            if key not in user_section:
                changes.append(f"Added {label} {key!r}")

    # Top-level scalar additions.
    for key in _TOP_LEVEL_SCALARS:
        if key not in user_json and key in shipped_json:
            changes.append(f"Set {key}")

    return result, changes


def format_diff_for_display(changes: list[str]) -> str:
    """Pretty-print a change list (one ``+ `` line per change)."""
    return "\n".join(f"  + {change}" for change in changes)


def parse_opencode_json(text: str) -> dict[str, Any]:
    """Parse an ``opencode.json`` document.

    Wraps :func:`json.loads` and re-raises parse errors as
    :class:`OpenCodeJsonInvalid` (exit code 3) with the original error
    included in the message.
    """
    try:
        parsed = json.loads(text)
    except json.JSONDecodeError as err:
        raise OpenCodeJsonInvalid(
            f"opencode.json is not valid JSON: {err.msg} (line {err.lineno}, column {err.colno})"
        ) from err
    if not isinstance(parsed, dict):
        raise OpenCodeJsonInvalid(
            f"opencode.json must be a JSON object, got {type(parsed).__name__}"
        )
    return parsed


def serialize_opencode_json(obj: dict[str, Any]) -> str:
    """Canonical serialization for stable on-disk diffs.

    ``json.dumps`` with ``indent=2`` and ``sort_keys=True`` plus a trailing
    newline so subsequent ``init`` runs produce byte-identical output.
    """
    return json.dumps(obj, indent=2, sort_keys=True) + "\n"

File: overlay/packages/opencode/src/neuralgentics/merge.ts
/**
 * `opencode.json` merge algorithm — TypeScript port of the Python
 * `neuralgentics-cli/src/neuralgentics/merge.py`.
 *
 * Deep merge with user-preservation semantics: the user's existing
 * `opencode.json` is the base, the shipped config is the overlay, and the
 * algorithm only ever ADDS missing entries — it never removes, overwrites
 * existing dicts, or touches the `provider` block.
 *
 * Rules (mirrors §4.1 of the design doc):
 *   1. `plugin` array → union (dedup, case-sensitive).
 *   2. `instructions` array → union (dedup).
 *   3. `provider` → preserved entirely from the user.
 *   4. `mcp` / `lsp` / `formatter` → add shipped keys missing in user.
 *   5. Top-level scalars (`$schema`, `autoupdate`, `tool_output`,
 *      `compaction`, `small_model`) → add from shipped if missing.
 */

/** The plugin entry added to the `plugin` array. */
export const PLUGIN_REFERENCE = "@veedubin/neuralgentics";

/** The instructions entry added to the `instructions` array. */
export const INSTRUCTIONS_REFERENCE = "AGENTS.md";

/** Top-level scalar keys added from the shipped config when missing. */
const TOP_LEVEL_SCALARS = [
  "$schema",
  "autoupdate",
  "tool_output",
  "compaction",
  "small_model",
] as const;

/** Dict-valued sections merged key-by-key (add new keys only). */
const DICT_SECTIONS = ["mcp", "lsp", "formatter"] as const;

/** Error thrown when an `opencode.json` document fails to parse. */
export class OpenCodeJsonInvalid extends Error {
  readonly exitCode = 3;
  readonly remediation = "Fix the JSON syntax error manually, then re-run.";
  constructor(message: string) {
    super(message);
    this.name = "OpenCodeJsonInvalid";
  }
}

/** A JSON object (loose typing for merge internals). */
type JsonObj = Record<string, unknown>;

/**
 * Parse an `opencode.json` document. Wraps `JSON.parse` and re-raises parse
 * errors as `OpenCodeJsonInvalid` (exit code 3).
 */
export function parseOpencodeJson(text: string): JsonObj {
  let parsed: unknown;
  try {
    parsed = JSON.parse(text);
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    throw new OpenCodeJsonInvalid(`opencode.json is not valid JSON: ${msg}`);
  }
  if (parsed === null || typeof parsed !== "object" || Array.isArray(parsed)) {
    const got = parsed === null ? "null" : Array.isArray(parsed) ? "array" : typeof parsed;
    throw new OpenCodeJsonInvalid(
      `opencode.json must be a JSON object, got ${got}`,
    );
  }
  return parsed as JsonObj;
}

/**
 * Canonical serialization for stable on-disk diffs. `JSON.stringify` with
 * 2-space indent + sorted keys + trailing newline.
 */
export function serializeOpencodeJson(obj: JsonObj): string {
  return JSON.stringify(sortKeys(obj), null, 2) + "\n";
}

/** Recursively sort object keys for stable output (mirrors `sort_keys=True`). */
function sortKeys(value: unknown): unknown {
  if (Array.isArray(value)) {
    return value.map(sortKeys);
  }
  if (value === null || typeof value !== "object") {
    return value;
  }
  const obj = value as JsonObj;
  const out: JsonObj = {};
  for (const key of Object.keys(obj).sort()) {
    out[key] = sortKeys(obj[key]);
  }
  return out;
}

/**
 * Case-sensitive, order-preserving union of two arrays. User entries come
 * first (in their original order), followed by any shipped entries not
 * already present. Duplicates are dropped.
 */
function unionArrays(user: unknown[] | undefined, shipped: unknown[] | undefined): unknown[] {
  const result: unknown[] = [];
  const seen = new Set<unknown>();
  for (const source of [user, shipped]) {
    if (!source) continue;
    for (const item of source) {
      // Hashable membership check: primitives dedup naturally; objects/arrays
      // fall back to reference identity (rare in opencode.json arrays, which
      // are normally strings).
      if (typeof item === "object" && item !== null) {
        if (!result.some((r) => r === item)) {
          result.push(item);
        }
      } else if (!seen.has(item)) {
        seen.add(item);
        result.push(item);
      }
    }
  }
  return result;
}

/** Deep clone via structuredClone (Node 20+). */
function deepClone<T>(value: T): T {
  if (typeof structuredClone === "function") {
    return structuredClone(value);
  }
  return JSON.parse(JSON.stringify(value)) as T;
}

/**
 * Merge a user `opencode.json` with the shipped config. Returns a NEW object;
 * the inputs are never mutated.
 */
export function mergeOpencodeJson(userJson: JsonObj, shippedJson: JsonObj): JsonObj {
  const result: JsonObj = deepClone(userJson);

  // 1 & 2 — array sections.
  result["plugin"] = unionArrays(
    userJson["plugin"] as unknown[] | undefined,
    shippedJson["plugin"] as unknown[] | undefined,
  );
  result["instructions"] = unionArrays(
    userJson["instructions"] as unknown[] | undefined,
    shippedJson["instructions"] as unknown[] | undefined,
  );

  // 3 — provider: PRESERVE user's entirely when present. When the user has
  // no provider block, add the shipped one verbatim.
  if (!("provider" in result) && "provider" in shippedJson) {
    result["provider"] = deepClone(shippedJson["provider"]);
  }

  // 4 — dict sections: add shipped keys missing in user.
  for (const section of DICT_SECTIONS) {
    const shippedSection = shippedJson[section];
    if (shippedSection === null || typeof shippedSection !== "object" || Array.isArray(shippedSection)) {
      continue;
    }
    const shippedObj = shippedSection as JsonObj;
    const resultSection = (result[section] as JsonObj | undefined) ?? {};
    if (!(section in result)) {
      result[section] = resultSection;
    }
    for (const [key, value] of Object.entries(shippedObj)) {
      if (!(key in resultSection)) {
        resultSection[key] = deepClone(value);
      }
    }
    if (section in result) {
      (result as JsonObj)[section] = resultSection;
    }
  }

  // 5 — top-level scalars.
  for (const key of TOP_LEVEL_SCALARS) {
    if (!(key in result) && key in shippedJson) {
      result[key] = deepClone(shippedJson[key]);
    }
  }

  return result;
}

/**
 * Like `mergeOpencodeJson` but also returns a list of human-readable change
 * descriptions. The list is empty when nothing changed (idempotent re-run).
 */
export function mergeOpencodeJsonWithDiff(
  userJson: JsonObj,
  shippedJson: JsonObj,
): { merged: JsonObj; changes: string[] } {
  const merged = mergeOpencodeJson(userJson, shippedJson);
  const changes: string[] = [];

  // Plugin additions.
  const userPlugin = (userJson["plugin"] as unknown[] | undefined) ?? [];
  const shippedPlugin = (shippedJson["plugin"] as unknown[] | undefined) ?? [];
  for (const item of shippedPlugin) {
    if (!userPlugin.includes(item)) {
      changes.push(`Added ${JSON.stringify(item)} to plugin array`);
    }
  }

  // Instructions additions.
  const userInstr = (userJson["instructions"] as unknown[] | undefined) ?? [];
  const shippedInstr = (shippedJson["instructions"] as unknown[] | undefined) ?? [];
  for (const item of shippedInstr) {
    if (!userInstr.includes(item)) {
      changes.push(`Added ${JSON.stringify(item)} to instructions array`);
    }
  }

  // Dict-section additions.
  const sectionLabels: Record<string, string> = {
    mcp: "MCP server",
    lsp: "LSP server",
    formatter: "formatter",
  };
  for (const [section, label] of Object.entries(sectionLabels)) {
    const shippedSection = shippedJson[section];
    if (shippedSection === null || typeof shippedSection !== "object" || Array.isArray(shippedSection)) {
      continue;
    }
    const shippedObj = shippedSection as JsonObj;
    const userSection = userJson[section];
    const userObj =
      userSection !== null && typeof userSection === "object" && !Array.isArray(userSection)
        ? (userSection as JsonObj)
        : {};
    for (const key of Object.keys(shippedObj)) {
      if (!(key in userObj)) {
        changes.push(`Added ${label} ${JSON.stringify(key)}`);
      }
    }
  }

  // Top-level scalar additions.
  for (const key of TOP_LEVEL_SCALARS) {
    if (!(key in userJson) && key in shippedJson) {
      changes.push(`Set ${key}`);
    }
  }

  return { merged, changes };
}

/** Pretty-print a change list (one `+ ` line per change). */
export function formatDiffForDisplay(changes: string[]): string {
  return changes.map((c) => `  + ${c}`).join("\n");
}